File: Graph_Algorithms/Shortest Path Algos/Dijkshtra's Algorithm/code.cpp
#include <bits/stdc++.h>
using namespace std;

class Graph
{
    int n;
    vector<list<pair<int, int>>> m;

public:
    Graph(int V)
    {
        n = V;
        m.resize(n);
    }

    void addEdge(int u, int v, int w = 1)
    {
        m[u].push_back({v, w});
        m[v].push_back({u, w});
    }

    void printAdjList()
    {
        for (auto i : m)
        {
            for (auto nbr : i)
            {
                cout << nbr.first << '(' << nbr.second << ") ";
            }
            cout << endl;
        }
    }

    int dijkshtras(int s, int dest)
    {
        set<pair<int, int>> q;
        vector<int> dist(n, INT_MAX);
        dist[s] = 0;
        q.insert({dist[s], s});
        while (!q.empty())
        {
            int u = q.begin()->second;
            cout << u << " ";
            q.erase(q.begin());
            for (auto nbr : m[u])
            {
                int v = nbr.first;
                int w = nbr.second;
                if (dist[v] > dist[u] + w)
                {
                    if (q.find({dist[v], v}) != q.end())
                        q.erase(q.find({dist[v], v}));
                    dist[v] = dist[u] + w;
                    q.insert({dist[v], v});
                }
            }
        }
        return dist[dest];
    }
};

int main()
{
    Graph g(5);
    g.addEdge(0, 1, 1);
    g.addEdge(1, 2, 2);
    g.addEdge(0, 2, 4);
    g.addEdge(0, 3, 7);
    g.addEdge(3, 2, 2);
    g.addEdge(3, 4, 3);
    g.printAdjList();
    cout << g.dijkshtras(0, 4) << endl;
    return 0;
}

File: Graph_Algorithms/Shortest Path Algos/Dijkshtra's Algorithm/code.py
from priority_queue import Minheap

class Graph:
    def __init__(self, len):
        self.V = len
        self.graph = [[] for _ in range(len)]

    def printAdjList(self):
        for i in range(len(self.graph)):
            print(i, end=' : ')
            print(self.graph[i])
        print("\n")
        
    def addEdge(self,u,v,w, isUndir = False):
        self.graph[u].append((v,w))
        if isUndir:
            self.graph[v].append((u,w))

    def dijkshtras(self,s,dest):
        n = len(self.graph)
        queue = Minheap(n)
        list = [0]*len(self.graph)
        dist = [float('inf') for i in range(n)]
        dist[s] = 0
        queue.enqueue((dist[s], s))
        while(queue.isEmpty() != 1):
            [a,b] = queue.dequeue()
            for i in self.graph[b]:
                x = i[0]
                w = i[1]    
                if dist[x] > dist[b] + w:
                    dist[x] = dist[b] + w
                    queue.enqueue((dist[x],x))
                    list[x] = b
        return list, dist[dest]

    def printPath(self,list,dest):
        if dest not in list:
            return None
        v = dest
        path = []
        while v != -1:
            path.append(v)
            v = list[v]
            list[v] = -1
        return path[::-1]

if __name__ == "__main__":
    g = Graph(4)
    g.addEdge(0,1,10)
    g.addEdge(0,3,5)
    g.addEdge(3,2,6)
    g.addEdge(2,1,8)
    g.addEdge(3,1,1)
    g.addEdge(1,3,7)
    g.printAdjList()
    [lst,d] = g.dijkshtras(0,3)
    print("Shortest Path is : ",d)
    print("Path : ",g.printPath(lst,3))
    
